jaxrs-client/src/main/java/io/micronaut/jaxrs/client/JaxRsInvocation.java

```diff
@@ -0,0 +1,525 @@
+import jakarta.ws.rs.core.GenericType;
+import jakarta.ws.rs.core.Response;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.net.URI;
+import java.util.List;
+import java.util.Locale;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.function.BiConsumer;
+
+/**
+ * The implementation of {@link Invocation}, {@link CompletionStageRxInvoker} and {@link AsyncInvoker}.
+ *
+ * @author Denis Stepanov
+ * @since 4.6
+ */
+@Internal
+final class JaxRsInvocation implements Invocation, CompletionStageRxInvoker, AsyncInvoker {
+
+    @NonNull
+    private final JaxRsClient client;
+    @NonNull
+    private final URI uri;
+    @NonNull
+    private final String method;
+    @Nullable
+    private final Entity<?> entity;
+
+    private final MutableHttpHeaders mutableHttpHeaders;
+
+    private final JaxRsConfiguration configuration;
+
+    JaxRsInvocation(JaxRsClient client,
+                    @NonNull URI uri,
+                    String method,
+                    @Nullable Entity<?> entity,
+                    MutableHttpHeaders mutableHttpHeaders,
+                    JaxRsConfiguration configuration) {
+        this.client = client;
+        this.uri = uri;
+        this.method = method;
+        this.entity = entity;
+        this.mutableHttpHeaders = mutableHttpHeaders;
+        this.configuration = configuration.copy();
+    }
+
+    @Override
+    public Invocation property(String name, Object value) {
+        configuration.addProperty(name, value);
+        return this;
+    }
+
+    @Override
+    public Response invoke() {
+        return asyncBlock(async(Argument.of(Response.class)));
+    }
+
+    @Override
+    public <T> T invoke(Class<T> aClass) {
+        return asyncBlock(async(Argument.of(aClass)));
+    }
+
+    @Override
+    public <T> T invoke(GenericType<T> genericType) {
+        return invoke(ArgumentUtil.from(genericType));
+    }
+
+    @Override
+    public Future<Response> submit() {
+        return async(Argument.of(Response.class));
+    }
+
+    @Override
+    public <T> Future<T> submit(Class<T> aClass) {
+        return async(Argument.of(aClass));
+    }
+
+    @Override
+    public <T> Future<T> submit(GenericType<T> genericType) {
+        return async(ArgumentUtil.from(genericType));
+    }
+
+    @Override
+    public <T> Future<T> submit(InvocationCallback<T> invocationCallback) {
+        return async(ArgumentUtil.from(invocationCallback)).whenComplete(withCallback(invocationCallback));
+    }
+
+    private <T> T invoke(Argument<T> type) {
+        return asyncBlock(async(type));
+    }
+
+    private <T> T asyncBlock(CompletableFuture<T> future) {
+        // Use it as a workaround before https://github.com/micronaut-projects/micronaut-core/pull/10975 is merged
```

Comment:
still relevant?

Reply:
No, but I will keep it this way to reuse the code for async/sync